feat: add rename when clicking on selected label page

LouisDelbosc opened the pull request:
PR for [issue#235](https://github.com/gristlabs/grist-core/issues/235)

Here is a demo :

https://user-images.githubusercontent.com/11460746/182617018-1faac739-4c47-43c8-b243-9c518c84621e.mov

If you have any comments or suggestions for improvement, they are welcome ! 🥳

berhalak reviewed (commented):
Works great. I think this is a very useful feature. One suggestion to discuss in the comments.

berhalak commented:
I've added some tests for the left panel menu that are already on the main branch, so you can rebase now and check if everything works. Feel free to add your own tests for this feature. Relevant tests are in `test/nbrowser/Pages.ts`

LouisDelbosc commented on app/client/ui2018/pages.ts:
```diff
@@ -69,7 +74,11 @@ export function buildPageDom(name: Observable<string>, actions: PageActions, ...
+            cssPageName(
+              dom.text(name),
+              testId('label'),
+              dom.on('click', (ev) => isTargetSelected(ev.target as HTMLElement) && isRenaming.set(true)),
```
You mean removing the check `isTargetSelected` and only go for double click or both ?

berhalak replied:
I think the flow with double click is enough, what do you think?

yohanboniface replied:
I think the "select + one click to rename" pattern was to be consistent with how it works with renaming column (cf #235 )

anaisconce replied:
Imagining it from a UX perspective, I like "select + one click to rename" better.

berhalak replied:
Ok, that makes sense. So I would only suggest making the label wider (flex-grow: 1), so that it is easier to click.

LouisDelbosc commented:
> I've added some tests for the left panel menu that are already on the main branch, so you can rebase now and check if everything works. Feel free to add your own tests for this feature. Relevant tests are in `test/nbrowser/Pages.ts`

I added a test next to the renaming tests you did. Is it fine for you ? Is there any other tests you thinks I should make ?

berhalak commented:
> > I've added some tests for the left panel menu that are already on the main branch, so you can rebase now and check if everything works. Feel free to add your own tests for this feature. Relevant tests are in `test/nbrowser/Pages.ts`
> 
> I added a test next to the renaming tests you did. Is it fine for you ? Is there any other tests you thinks I should make ?

The test look ok, thank you! Can you take a look also on the failing test? I think it needs some updates where we click on the page menu item, maybe it is enough to make it click on the page initials?

berhalak reviewed (changes requested):
The code looks ok, thank you.  Only the failing test needs to be updated.

LouisDelbosc commented:
> The code looks ok, thank you. Only the failing test needs to be updated.

I'm on it. I'll try to update it during the morning.

LouisDelbosc commented:
@berhalak Sorry some of my tests are not workings exactly like on CI so I thought it would work but it seems there was another test which was failing.

paulfitz commented:
@LouisDelbosc it might be useful to run tests under docker also. You can do that by building an image from your branch (with `docker build -t gristlabs/grist .`) and then using `test/test_under_docker.sh`. Just trying that now, I could quickly see the errors from CI by replacing the line `mocha _build/test/nbrowser/*.js "$@"` with `node_modules/.bin/mocha _build/test/nbrowser/Pages.js -b` in that script. Just an idea.

berhalak commented on test/nbrowser/Pages.ts:
```diff
@@ -126,6 +144,7 @@ describe('Pages', function() {
 
   it('should not change page when clicking the input while renaming page', async () => {
     // check that initially People is selected
+    await driver.findContent('.test-treeview-itemHeader', 'People').doClick();
```
I think this is the problem. I suggest adding a little helper at the end:

```
function clickPage(name: string|RegExp) {
  return driver.findContent('.test-treeview-itemHeader', name).find(".test-docpage-initial").doClick();
}
```

And replacing all lines that want to change the page with a call to that helper.
Tests are sensitive to OS that they are run at. I think on OSX and Windows they click at the top left corner, while on linux they click at the center (or other way around :) ).
That is why you see the difference, as they clicking on the label instead of the initials.

berhalak reviewed (approved):
Thanks, works great.
Now I think we should handle the Escape key in the same way as for columns (to cancel the change), but it can be a follow-up task.

paulfitz commented:
Thanks @LouisDelbosc ! Your improvement should make its way into Grist docker images tomorrow, and to hosted Grist by next Monday.

paulfitz commented:
I'm enjoying using this on my own docs on https://docs.getgrist.com, thanks again @LouisDelbosc :-)

anaisconce commented:
What a treat to come back to this after a week off! I used it immediately, love the experience. Thank you so much @LouisDelbosc !
